onnxruntime/core/providers/cuda/tensor/expand.cc

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "expand.h"
+#include "expand_impl.h"
+#include "core/providers/cpu/tensor/utils.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+#define REGISTER_KERNEL_TYPED(T)                                 \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                 \
+      Expand,                                                    \
+      kOnnxDomain,                                               \
+      8,                                                         \
+      T,                                                         \
+      kCudaExecutionProvider,                                    \
+      KernelDefBuilder()                                         \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>()) \
+          .InputMemoryType<OrtMemTypeCPUInput>(1),               \
+      Expand<T>);
```

Comment:
Please consider implementing typeless kernel for tensor ops. In the case of Expand, only size of element matters, so you just need uin8/uint16/uint32/uint64 in the actual CUDA kernels, and handle type size in Compute.

Reply:
Handling types using a case statement makes the code look quite messy (especially for CUDA operators). Is there a good/elegant example you can point to?

Reply:
Well, the same could be said for the typed kernels in cuda_execution_provider.cc. You may take a look at Slice op as reference.

---
In reply to: [297457252](https://github.com/microsoft/onnxruntime/pull/1292#discussion_r297457252) [](ancestors = 297457252)

Reply:
Reduced the number of types to just the 4 mentioned above.